### PR TITLE
adding KMeans PyTorch Implementation to cfa model

### PR DESCRIPTION
**Pull request by aadhamm**
# Description

- Provide an implementation to the KMeans clustering algorithm through the PyTorch framework

- Fixes # (956)

## Changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which refactors the code base)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist

- [x] My code follows the [pre-commit style and check guidelines](https://github.com/openvinotoolkit/anomalib/blob/main/CONTRIBUTING.md) of this project.
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing tests pass locally with my changes
- [ ] I have added a summary of my changes to the [CHANGELOG](https://github.com/openvinotoolkit/anomalib/blob/main/CHANGELOG.md) (not for minor changes, docs and tests).

### CODE REVIEW

File: src/anomalib/models/cfa/torch_model.py [hunk 1]
```diff
@@ -78,6 +78,66 @@ def get_feature_extractor(backbone: str, return_nodes: list[str]) -> GraphModule
...
+
+            # Update the centroids to be the mean of the data points assigned to them
+            for j in range(self.n_clusters):
+                mask = self.cluster_assignments == j
+                if mask.any():
+                    self.centroids[j] = X[mask].mean(dim=0)
+                    
+        #thise line returns labels and centoids of the results, 
+        #alternative to Sklearn's cluster_centers_ & labels_ attributes            
+        return self.cluster_assignments, self.centroids
```
Comment (samet-akcay) [inline src/anomalib/models/cfa/torch_model.py:125]: Would it be possible to return `self`, similar to the `scikit-learn` implementation?

**Review by samet-akcay — CHANGES_REQUESTED** [review]
Thanks for creating this. I wanted to compare the scikit-learn and torch implementations, but had some trouble. Can you share the steps that you mentioned here in this link
https://github.com/openvinotoolkit/anomalib/issues/956#issuecomment-1493289548

**Comment by aadhamm** [comment]
Dear Samet, 

thank you for putting such effort into review and feedback, 

of course, all of your requested changes are possible and I will start working on them and commit changes ASAP, I will also provide you with the full code/notebook I used for comparison between the two implementations.

in the meantime, I potentially looking to address some other TO-DO issues either were the other two issues in cfa model or issues in other models as well, if that's possible and eligible of course.

Thank you.

**Comment by aadhamm** [comment]
Dear @samet-akcay, 

please forgive me for the late reply, due to several engagements in the last few days, 

here is my full approach to comparing the Kmeans Sklean and PyTorch implementations through silhouette score, which is a metric used to calculate the goodness of a clustering technique. Its value ranges from -1 to 1. 

It contains the modifications you asked for, this is a preview, I will commit the changes as required and I may add a running notebook with the following comparison code, if it's not necessary, I will delete it.

Thank you



```
# Import necessary libraries
import torch
from sklearn.datasets import make_blobs
from sklearn.cluster import KMeans
import numpy as np
from sklearn.metrics import silhouette_score
##sklearn KMeans is imported lately so it's class not be confused with pytorch class

class KMeans:
    def __init__(self, n_clusters: int, max_iter:int = 10):
        
        """
        Initializes the KMeans object.

        Parameters:
            n_clusters: The number of clusters to create.
            max_iter: The maximum number of iterations to run the algorithm for.
        """
        self.n_clusters = n_clusters
        self.max_iter = max_iter

    def fit(self, X):
        """
        Runs the k-means algorithm on input data X.

        Parameters:
            X: A tensor of shape (N, D) containing the input data.
            N is the number of data points 
            D is the dimensionality of the data points.
        """
        N, D = X.shape

        # Initialize centroids randomly from the data points
        centroid_indices = torch.randint(0, N, (self.n_clusters,))
        self.cluster_centers_ = X[centroid_indices]

        # Run the k-means algorithm for max_iter iterations
        for i in range(self.max_iter):
            # Compute the distance between each data point and each centroid
            distances = torch.cdist(X, self.cluster_centers_)

            # Assign each data point to the closest centroid
            self.labels_ = torch.argmin(distances, dim=1)

            # Update the centroids to be the mean of the data points assigned to them
            for j in range(self.n_clusters):
                mask = self.labels_ == j
                if mask.any():
                    self.cluster_centers_[j] = X[mask].mean(dim=0)
                    
        #thise line returns labels and centoids of the results,          
        return self.labels_, self.cluster_centers_

    def predict(self, X):
        """
        Assigns each data point in X to its closest centroid.

        Parameters:
            X: A tensor of shape (N, D) containing the input data.

        Returns:
            A tensor of shape (N,) containing the index of the closest centroid for each data point.
        """
        distances = torch.cdist(X, self.cluster_centers_)
        return torch.argmin(distances, dim=1)



# Generate sample data
X, _ = make_blobs(n_samples=1000, centers=5, n_features=10, random_state=42)

X_torch = torch.tensor(X, dtype=torch.float32)


# Set parameters
n_clusters = 3
max_iter = 3000

# Run KMeans using Torch implementation
kmeans_torch = KMeans(n_clusters=n_clusters, max_iter=max_iter)
kmeans_torch.fit(X_torch)


from sklearn.cluster import KMeans
# Run KMeans using scikit-learn implementation
kmeans_sklearn = KMeans(n_clusters=n_clusters, max_iter=max_iter)
kmeans_sklearn.fit(X)



sklearn_labels = kmeans_sklearn.labels_
pytorch_labels = kmeans_torch.labels_

sklearn_centers = kmeans_sklearn.cluster_centers_
pytorch_centers = kmeans_torch.cluster_centers_



# Check that the results are the same
print("Scikit-learn centroids:\n", sklearn_centers)
print("Torch centroids:\n", pytorch_centers)

print("Scikit-learn labels:\n", sklearn_labels)
print("Torch labels:\n", pytorch_labels) 



import numpy as np
from sklearn.metrics import silhouette_score


# Calculate the silhouette score
silhouette_sklearn = silhouette_score(X, sklearn_labels)

# Calculate the silhouette score
silhouette_torch = silhouette_score(X_torch, pytorch_labels)


# Print the comparison results
print(f"Silhouette score for Sklearn: {silhouette_sklearn}")
print(f"Silhouette score for Pytorch: {silhouette_torch}")

```

**Comment by aadhamm** [comment]
@samet-akcay, I wanted to ask if this to-do task, `TODO: Replace this with the new torchfx feature extractor` has already been done, because I am willing to continue with the rest of them in the same CFA implementation. 

As far as I know, `create_feature_extractor ()` is the new fx feature extractor.

**Comment by aadhamm** [comment]
@samet-akcay, if it's already been done, I may contribute to this task, `# TODO: Automatically infer the number of dims`.

**Comment by samet-akcay** [comment]
> @samet-akcay, I wanted to ask if this to-do task, `TODO: Replace this with the new torchfx feature extractor` has already been done? because I am willing to continue with the rest of them in the same CFA implementation.
> 
> As far as I know, `create_feature_extractor()` is the new fx feature extractor.

@ashwinvaidya17, can you help here to show how @aadhamm can utilize the new feature extractor for the cfa model

**Comment by ashwinvaidya17** [comment]
> @samet-akcay, I wanted to ask if this to-do task, `TODO: Replace this with the new torchfx feature extractor` has already been done? because I am willing to continue with the rest of them in the same CFA implementation.
> 
> As far as I know, `create_feature_extractor()` is the new fx feature extractor.

You can have a look here to see how it is done. You can adapt your code to call this class. https://github.com/openvinotoolkit/anomalib/blob/4fe6c7457354f18d95f64fc4f81775836628c9ba/src/anomalib/models/csflow/torch_model.py#L475-L476
If you still have any questions then feel free to ask

**Review by ashwinvaidya17 — CHANGES_REQUESTED** [review]
Thanks for the efforts I have a few comments.
- Can you remove the jupyter notebook from the components folder. If you want to confirm the outputs then you can move this code into a file and turn it into a unit test.
- Can you also address the issues raised by codacy.
- The spinx parser is configured to follow the Google's docstring format. Can you update the docstrings to conform to this format? Here is an example https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html

**Comment by aadhamm** [comment]
@ashwinvaidya17 Thank you for your detailed review, I will work on all the required changes and update you as soon as possible and yes, you are right, there are variables that could be reduced as it will work best for codacy tests.

**Comment by aadhamm** [comment]
@samet-akcay @ashwinvaidya17 Hello everyone, I have made the required changes and addressed issues by codacy.

If you could take a look and give me your feedback I will appreciate it a lot.

Thank you.

File: src/anomalib/models/components/cluster/kmeans.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        self.max_iter = max_iter
+
+    def fit(self, X):
+        """
+        Runs the k-means algorithm on input data X.
+
+        Parameters:
+            X: A tensor of shape (N, D) containing the input data.
+            N is the number of data points 
+            D is the dimensionality of the data points.
```
[the remark this answers is not in the document]
Comment (aadhamm) [reply src/anomalib/models/components/cluster/kmeans.py]: according to more than one source, N is the number of samples and D is the number of features, D is s used to initialize the centroids randomly with the same dimensionality as the input data, but not used actually afterward in the code which created issues related to codacy tests.

I resolved this later using your suggestion.


[K-means clustering - PyTorch API](https://www.kernel-operations.io/keops/_auto_tutorials/kmeans/plot_kmeans_torch.html)
[Fast Pytorch Kmeans
](https://github.com/DeMoriarty/fast_pytorch_kmeans/blob/master/fast_pytorch_kmeans/kmeans.py)

**Comment by aadhamm** [comment]
@samet-akcay @ashwinvaidya17 Hello, everyone i have done the required changes and all the checks have passed, if you could take a look and see if it is ready to merge or not yet.

Thank you.

**Review by ashwinvaidya17 — APPROVED** [review]
Sorry for not getting back with this. Thanks for all the efforts.

**Review by ashwinvaidya17 — APPROVED** [review]
